File: tests/unit-tests/config_tests_rust/deterministic/dnstest.rs
use std::net::{UdpSocket, Ipv4Addr, SocketAddr};
use std::io::{self, Write};

// DNS Header structure
#[repr(C)]
struct DnsHeader {
    xid: u16,
    flags: u16,
    qdcount: u16,
    ancount: u16,
    nscount: u16,
    arcount: u16,
}

// DNS Record A structure
#[repr(C)]
struct DnsRecordAT {
    compression: u16,
    typ: u16,
    clas: u16,
    ttl: u32,
    length: u16,
    addr: u32, // IPv4 address as u32
}

fn main() -> io::Result<()> {
    // Create UDP socket
    let socket = UdpSocket::bind("0.0.0.0:0")?;
    println!("is sockfd valid? yes");

    // Create DNS header
    let mut dnsh = DnsHeader {
        xid: 0x1234,
        flags: 0x0100,
        qdcount: 0x0001,
        ancount: 0,
        nscount: 0,
        arcount: 0,
    };

    // Hostname for DNS query: engineering.nyu.edu
    let hostname = b"\x0bengineering\x03nyu\x03edu\x00";
    let dnstype: u16 = 1; // A record
    let dnsclass: u16 = 1; // IN class

    // Construct packet
    let packet_len = std::mem::size_of::<DnsHeader>() + hostname.len() + 2 + 2;
    let mut packet = vec![0u8; packet_len];
    
    // Copy DNS header
    let header_bytes = unsafe {
        std::slice::from_raw_parts(
            &dnsh as *const DnsHeader as *const u8,
            std::mem::size_of::<DnsHeader>()
        )
    };
    packet[..header_bytes.len()].copy_from_slice(header_bytes);
    
    // Copy hostname
    let hostname_start = std::mem::size_of::<DnsHeader>();
    packet[hostname_start..hostname_start + hostname.len()].copy_from_slice(hostname);
    
    // Copy query type and class
    let type_start = hostname_start + hostname.len();
    packet[type_start..type_start + 2].copy_from_slice(&dnstype.to_be_bytes());
    packet[type_start + 2..type_start + 4].copy_from_slice(&dnsclass.to_be_bytes());

    // Set up destination address (OpenDNS: 208.67.222.222:53)
    let dest_addr = SocketAddr::new(
        Ipv4Addr::new(208, 67, 222, 222).into(),
        53
    );

    // Send DNS query
    let sent_len = socket.send_to(&packet, dest_addr)?;
    println!("{} is the length sent", sent_len);

    // Receive DNS response
    let mut dns_resp = [0u8; 512];
    let (received_len, _) = socket.recv_from(&mut dns_resp)?;
    println!("{} is the length received", received_len);

    // Extract response header
    let response_header = unsafe {
        &*(dns_resp.as_ptr() as *const DnsHeader)
    };

    // Check if response is valid (no error flags)
    if (response_header.flags.to_be() & 0xf) != 0 {
        println!("DNS response contains error flags");
        return Ok(());
    }

    // Skip over the name in response
    let mut name_ptr = std::mem::size_of::<DnsHeader>();
    while name_ptr < received_len && dns_resp[name_ptr] != 0 {
        name_ptr += dns_resp[name_ptr] as usize + 1;
    }

    // Skip null byte, qtype, and qclass to get to answer section
    let record_ptr = name_ptr + 5;
    if record_ptr + std::mem::size_of::<DnsRecordAT>() <= received_len {
        let record = unsafe {
            &*(dns_resp.as_ptr().add(record_ptr) as *const DnsRecordAT)
        };
        
        // Convert IPv4 address to string
        let ip_bytes = record.addr.to_be_bytes();
        let ip_addr = Ipv4Addr::new(ip_bytes[0], ip_bytes[1], ip_bytes[2], ip_bytes[3]);
        println!("the dns lookup yields the ip address: {}", ip_addr);
    } else {
        println!("DNS response too short to extract answer");
    }

    Ok(())
} 

File: tests/unit-tests/networking_tests/deterministic/dnstest.c
#include <arpa/inet.h>
#include <assert.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>

int main(void) {
    struct DnsHeader {
        uint16_t xid;
        uint16_t flags;
        uint16_t qdcount;
        uint16_t ancount;
        uint16_t nscount;
        uint16_t arcount;
    };

    struct DnsRecordAT {
        uint16_t compression;
        uint16_t typ;
        uint16_t clas;
        uint32_t ttl;
        uint16_t length;
        struct in_addr addr;
    } __attribute__((packed));

    int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    assert(sockfd >= 0);

    char *hostname = "\13engineering\03nyu\03edu\0";
    uint16_t dnstype = htons(1);
    uint16_t dnsclass = htons(1);

    int packetlen = sizeof(struct DnsHeader) + (int)strlen(hostname) + 1 + 2 + 2;
    char packvla[packetlen];
    int packetlenleft = packetlen;

    struct DnsHeader dnsh;
    dnsh.xid = htons(0x1234);
    dnsh.flags = htons(0x0100);
    dnsh.qdcount = htons(0x0001);
    dnsh.ancount = 0;
    dnsh.nscount = 0;
    dnsh.arcount = 0;

    memcpy(packvla, &dnsh, sizeof(struct DnsHeader));
    packetlenleft -= sizeof(struct DnsHeader);
    memcpy(packvla + packetlen - packetlenleft, hostname, strlen(hostname) + 1);
    packetlenleft -= (int)strlen(hostname) + 1;
    memcpy(packvla + packetlen - packetlenleft, &dnstype, 2);
    packetlenleft -= 2;
    memcpy(packvla + packetlen - packetlenleft, &dnsclass, 2);
    packetlenleft -= 2;
    assert(packetlenleft == 0);

    /* Canned DNS response: header + question (same name) + one A record 192.0.2.1 */
    static const unsigned char canned[] = {
        /* Header: xid=0x1234, flags=0x8180, qd=1, an=1, ns=0, ar=0 */
        0x12, 0x34, 0x81, 0x80, 0x00, 0x01, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00,
        /* Question: \x0b engineering \x03 nyu \x03 edu \x00 */
        0x0b, 'e','n','g','i','n','e','e','r','i','n','g', 0x03, 'n','y','u', 0x03, 'e','d','u', 0x00,
        /* qtype=1, qclass=1 */
        0x00, 0x01, 0x00, 0x01,
        /* Answer: ptr to 12, type A, class IN, ttl=60, rdlength=4, 192.0.2.1 */
        0xc0, 0x0c, 0x00, 0x01, 0x00, 0x01, 0x00, 0x00, 0x00, 0x3c, 0x00, 0x04,
        192, 0, 2, 1
    };

    const char *dnsresp = (const char *)canned;
    const struct DnsHeader *rdnsh = (const struct DnsHeader *)dnsresp;

    assert(!(ntohs(rdnsh->flags) & 0xf));

    const char *nameptr = dnsresp + sizeof(struct DnsHeader);
    while (*nameptr != 0) {
        nameptr += (unsigned char)*nameptr + 1;
    }

    const struct DnsRecordAT *recordptr = (const struct DnsRecordAT *)(nameptr + 5);
    assert(ntohs(recordptr->typ) == 1);
    assert(ntohs(recordptr->clas) == 1);
    assert(ntohs(recordptr->length) == 4);
    {
        const unsigned char *a = (const unsigned char *)&recordptr->addr;
        assert(a[0] == 192 && a[1] == 0 && a[2] == 2 && a[3] == 1);
    }

    printf("dnstest ok\n");
    return 0;
}
